Automated Docker image build and test

caisq opened the pull request:
1) Breaking up the pip.sh into the install and test-on-install

This is aimed at automating testing of docker images.
Also, a step is added in pip.sh to uninstall existing versions of
protobuf and tensorflow if they exist. This addresses pip install-test
issues on non-Docker environments (e.g., Mac)

2) Using virtualenv to perform pip test-on-install

The built pip-package is installed in a virtualenv, then the pip install test and the tutorial
test are carried out inside the virutal env. This is done inside the same docker container as the one used for bazel build, if docker is available. 

3) Automating Docker image build and test

ci_parameterized_build now has a new env-var parameter:
TF_BUILD_DOCKER_TEST, that can be used to trigger Docker image build
and test following the PIP build (i.e., TF_BUILD_IS_PIP=PIP or BOTH).

The image is automatically tagged as ${USER}/tensorflow:${VERSION},
wherein the version is extracted from version.h.

The main script for Docker build and test is the newly added:
docker_test.sh

It uses Dockerfiles in tools/ci_build/docker and installs the whl files
inside the Docker image. After the pip installation step, the Python
unit tests and tutorial tests are inside the Docker container.

caisq committed:
Automated Docker image build and test

1) Breaking up the pip.sh into the install and test-on-install

This is aimed at automating testing of docker images.
Also, a step is added in pip.sh to uninstall existing versions of
protobuf and tensorflow if they exist. This addresses pip install-test
issues on non-Docker environments (e.g., Mac)

2) Automating Docker image build and test

ci_parameterized_build now has a new env-var parameter:
TF_BUILD_DOCKER_TEST, that can be used to trigger Docker image build
and test following the PIP build (i.e., TF_BUILD_IS_PIP=PIP or BOTH).

The image is automatically tagged as ${USER}/tensorflow:${VERSION},
wherein the version is extracted from version.h.

The main script for Docker build and test is the newly added:
docker_test.sh

It uses Dockerfiles in tools/ci_build/docker and installs the whl files
inside the Docker image. After the pip installation step, the Python
unit tests and tutorial tests are inside the Docker container.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

jendap commented on tensorflow/tools/docker/Dockerfile:
```diff
@@ -29,8 +30,8 @@ RUN pip --no-cache-dir install \
 ENV TENSORFLOW_VERSION 0.7.1
-RUN pip --no-cache-dir install \
-    http://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-${TENSORFLOW_VERSION}-cp27-none-linux_x86_64.whl
+ADD whl/tensorflow-*.whl /tmp/
```
The problem with doing this is that people can simply build docker in tools/docker folder anymore. They would have to know they have to run multiple hops of other scripts. Complicated.

caisq replied:
Unfortunately it's difficult to parameterize Dockerfiles. How about this, I can move this content into Dockerfile.devel (replacing the existing ones) and restore the http URL in this file?

jendap replied:
It would be nice if all the docker files just work. But we do not need this do we? We will use virtualenv for normal builds anyway. We will want to test this only when we do a release, right? We can script that later and differently. We can simply build and upload the pips before building the docker images.

jendap commented on tensorflow/tools/docker/Dockerfile.gpu:
```diff
@@ -1,9 +1,10 @@
-FROM nvidia/cuda:7.5-cudnn4-runtime
+FROM nvidia/cuda:7.5-cudnn4-devel
```
This container should not depend on devel as it would make it a lot bigger. And this container is what we recommend people to use for runtime. It should be small.

caisq replied:
I changed it to devel because otherwise there will be a couple of test failures. Some tests depend on devel headers and libraries. Maybe I can just blacklist those tests here?

jendap replied:
Really? Which tests depends on it? I thought it runs python scripts. The devel stuff should be only needed if you compile some c/c++ code. And that should not be happening.

If it really is a problem for a few tests then I would ignore them here.

jendap commented on tensorflow/tools/ci_build/builds/pip_test.sh:
```diff
@@ -0,0 +1,314 @@
+#!/usr/bin/env bash
```
Nice. It is good to pull out the test only script. No compilation or installation here.

Should it be called test_installation.sh? Or something like that? I can be used to run python test on tf installed anywhere by any means.

jendap commented on tensorflow/tools/ci_build/builds/docker_test.sh:
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env bash
```
What is the goal of this? Are we trying to automate the release process? Or are we trying to run python tests for every build (master, pull requests, ...) in better isolation?

For the better isolation it would be better (simpler, multiplatform, easier and faster) to use virtualenv. We can make release more automated later.

caisq replied:
The main goal of this file (docker_test.sh) is to automate the docker image release process. I have now made further changes to use virtualenv.

jendap commented on tensorflow/tools/ci_build/builds/pip.sh:
```diff
@@ -72,12 +37,56 @@ abs_path() {
 }
 
+# Uninstall existing pip packages
+uninstall_existing_pip_packages() {
```
The uninstall sounds a bit dangerous. You would never need it if it was intended to run only in inside per build docker container. Because there would be nothing to uninstall in the first place. Also this does not allow concurrent builds outside docker (think mac os).

It would be great to think about any developer being able to use it on his/her computer. Without being afraid of messing up system. Hence no uninstall. Would it work if we would create per build virtualenv for python?

One can have working version of tensorflow installed on system and develop tensorflow on the same machine then.

caisq committed:
Replacing building inside Docker with virtualenv bulid-tests

Now the building are no longer done inside Docker containers. It is
carried out on the host for both Linux and Mac. The built pip-package is
installed in a virtualenv, then the pip install test and the tutorial
test are carried out inside the virutal env.

Of course, the Docker build-test are still done inside Docker containers.

caisq committed:
Chaser 1

caisq committed:
Chaser 2: Perfomring build with docker if available

vitualenv pip install-test is carried out inside a Docker container if
docker is availble.

caisq committed:
Merge branch 'master' into docker-dev-1

caisq committed:
Chaser: Minor fixes

caisq committed:
Chaser: Minor fix 2

caisq committed:
Merge branch 'master' into docker-dev-1

jendap commented on tensorflow/tools/ci_build/builds/pip.sh:
```diff
@@ -140,189 +125,47 @@ echo "whl file path = ${WHL_PATH}"
-${PYTHON_BIN_PATH} -m pip install -v --user ${WHL_PATH} \
+# Create temporary directory for install test
+VENV_DIR="${PIP_TEST_ROOT}/venv"
+rm -rf "${VENV_DIR}" && mkdir -p "${VENV_DIR}"
```
Does pip inside virtualenv work reliably without deleting the whole env? It would be nice for caching. We would reinstall only tf and dependencies only if their version would change. But I'm not sure if it would be reliable.

jendap commented on tensorflow/tools/ci_build/builds/pip_test.sh:
```diff
@@ -0,0 +1,297 @@
+"(Major.Minor version: ${PY_MAJOR_MINOR_VER})"
+
+# Avoid permission issues outside container
+umask 000
```
That's weird. What issues?

caisq replied:
When the docker container runs, the user is root. It turns out that directories created inside the container will retain root owner ship even outside the container.

jendap commented on tensorflow/tools/ci_build/ci_parameterized_build.sh:
```diff
@@ -50,9 +50,18 @@
+#                      tutorials tests (Applicable only if TF_BUILD_IS_PIP is
+#                      PIP or BOTH).
+#                      See builds/test_tutorials.sh
+#   TF_BUILD_DOCKER_TEST:
```
Do you really want to put the release docker testing into this script? There is already way too many things here, is it not?

jendap commented on tensorflow/tools/ci_build/install/install_deb_packages.sh:
```diff
@@ -38,3 +38,5 @@ apt-get install -y \
 apt-get clean
 rm -rf /var/lib/apt/lists/*
+
+pip install virtualenv
```
Would python-virtualenv from apt-get work? It would be better to use the linux distro version if possible.

jendap commented:
Splitting the test out of pip.sh is good. Using virtualenv is good. Testing the docker images is, imho, nice to have but not critical for now. We should not change the docker files. Also putting docker files testing into parametrized build is, imho, an overkill.

caisq committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into docker-dev-1

caisq committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into docker-dev-1

caisq committed:
Chaser 4: Using apt-get virutalenv

caisq committed:
Chaser 6: Going back to runtime gpu dockerfile

caisq committed:
Chaser 7: Reverting some chnages in Dockerfiles

caisq committed:
Chaser 8

caisq committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into docker-dev-1

caisq committed:
Chaser 9: Removing docker_test-related lines in
ci_parameterized_build.sh

caisq committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into docker-dev-1

caisq committed:
Chaser 10

caisq commented:
Addressed @jendap's comments. Closing this PR and opening a new fresh one with all changes squashed. https://github.com/tensorflow/tensorflow/pull/1380
